### PR TITLE
TVS2: New tiered-version-store implementation / non-transactional base

### PR DESCRIPTION
**Pull request by snazy**
Add the base class for all non-transactional / no-sql database adapters. The abstract class `NonTxDatabaseAdapter` contains the common logic for this kind of databases.

Actual implementations (e.g. in-memory, rocks-db, etc) then implement an extension of `DatabaseAdapterConfig` to supply e.g. database specific connection details, an implementation of `DatabaseAdapterFactory` and of course an implementation of `NonTxDatabaseAdapter` to implement the database specific read/write/cas operations.

None of the no-sql database I can think of require "paged" reads to happen against the same "borrowed connection", so `NonTxOperationContext` is a dummy singleton.

Note: concrete implementations (coming as follow-up PRs) are contained in [this branch](https://github.com/snazy/nessie/commits/tvs2-with-legacy) in the commits named `TVS2: New tiered-version-store / databaser-adapters / In-Memory` and `TVS2: New tiered-version-store / databaser-adapters / RocksDB`

---
This change is [Reviewable](https://reviewable.io/reviews/projectnessie/nessie/1786)

### CODE REVIEW

File: versioned/persist/nontx/src/main/java/org/projectnessie/versioned/persist/nontx/NonTxDatabaseAdapter.java [hunk 1]
```diff
@@ -0,0 +1,751 @@
...
+          ref,
+          false,
+          false,
+          (ctx, pointer, branchCommits, newKeyLists) -> {
+            if (pointer.getNamedReferencesMap().containsKey(ref.getName())) {
+              throw referenceAlreadyExists(ref);
+            }
+
+            Hash hash = target;
+            if (hash == null) {
```
Comment (nastra) [inline versioned/persist/nontx/src/main/java/org/projectnessie/versioned/persist/nontx/NonTxDatabaseAdapter.java:331]: maybe: `Hash hash = null == target ? NO_ANCESTOR : target`

File: versioned/persist/nontx/src/main/java/org/projectnessie/versioned/persist/nontx/NonTxDatabaseAdapter.java [hunk 2]
```diff
@@ -0,0 +1,751 @@
...
+    }
+  }
+
+  @Override
+  public void delete(NamedRef reference, Optional<Hash> expectedHead)
+      throws ReferenceNotFoundException, ReferenceConflictException {
+    try {
+      casOpLoop(
+          reference,
+          true,
```
[the remark this answers is not in the document]
Comment (nastra) [reply versioned/persist/nontx/src/main/java/org/projectnessie/versioned/persist/nontx/NonTxDatabaseAdapter.java:358]: could we figure out on-the-fly in `casOpLoop` whether it's a delete/commit operation maybe?

File: versioned/persist/nontx/src/main/java/org/projectnessie/versioned/persist/nontx/NonTxDatabaseAdapter.java [hunk 3]
```diff
@@ -0,0 +1,751 @@
...
+              Hash.of(e.getValue().getHash()));
+
+      result = Stream.concat(result, perRef);
+    }
+
+    return result;
+  }
+
+  // /////////////////////////////////////////////////////////////////////////////////////////////
+  // NMon-Transactional DatabaseAdapter subclass API (protected)
```
Comment (nastra) [inline versioned/persist/nontx/src/main/java/org/projectnessie/versioned/persist/nontx/NonTxDatabaseAdapter.java:494]: nit: typo

File: versioned/persist/nontx/src/main/java/org/projectnessie/versioned/persist/nontx/NonTxDatabaseAdapter.java [hunk 4]
```diff
@@ -0,0 +1,751 @@
...
+  /**
+   * Writes a global-state-log-entry without any operations, just to move the global-pointer
+   * forwards for a "proper" CAS operation.
+   */
+  // TODO maybe replace with a 2nd-ary value in global-state-pointer to prevent the empty
+  //  global-log-entry
+  protected Hash noopGlobalLogEntry(NonTxOperationContext ctx, GlobalStatePointer pointer)
+      throws ReferenceConflictException {
+    // Need a new empty global-log entry to be able to CAS
+    long timeInMicros = commitTimeInMicros();
```
Comment (nastra) [inline versioned/persist/nontx/src/main/java/org/projectnessie/versioned/persist/nontx/NonTxDatabaseAdapter.java:686]: nit: can be inlined

File: versioned/persist/nontx/src/main/java/org/projectnessie/versioned/persist/nontx/NonTxDatabaseAdapter.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (snazy) [reply versioned/persist/nontx/src/main/java/org/projectnessie/versioned/persist/nontx/NonTxDatabaseAdapter.java:358]: added an enum

File: versioned/persist/nontx/src/main/java/org/projectnessie/versioned/persist/nontx/NonTxDatabaseAdapter.java (hunk 1 above)
Comment (snazy) [reply versioned/persist/nontx/src/main/java/org/projectnessie/versioned/persist/nontx/NonTxDatabaseAdapter.java:331]: Nah - I'd like to keep the comment ther.

Comment (nastra) [reply versioned/persist/nontx/src/main/java/org/projectnessie/versioned/persist/nontx/NonTxDatabaseAdapter.java:293]: you can still keep the comment fwiw :)

**Review by rymurr — APPROVED** [review]

**Pull request merged**